src/ostree/ot-admin-builtin-unlock.c

```diff
@@ -0,0 +1,356 @@
+}
+
+static gboolean
+unlock_deployment (OstreeSysroot *sysroot,
+                   OstreeDeployment *target_deployment,
+                   const char *deployment_mode,
+                   GCancellable *cancellable,
+                   GError **error)
+{
+  gboolean ret = FALSE;
+  glnx_unref_object OstreeDeployment *deployment_clone =
+    ostree_deployment_clone (target_deployment);
+  GKeyFile *origin_clone = ostree_deployment_get_origin (deployment_clone);
+
+  g_key_file_set_string (origin_clone, "origin", "unlocked",
+                         deployment_mode);
+
+  if (!ostree_sysroot_write_origin_file (sysroot, target_deployment, origin_clone,
+                                         cancellable, error))
+    goto out;
+
+  if (!ostree_sysroot_deployment_set_mutable (sysroot, target_deployment, TRUE,
+                                              cancellable, error))
+    goto out;
+
+  ret = TRUE;
+ out:
+  return ret;
+}
+
+gboolean
+ot_admin_builtin_unlock (int argc, char **argv, GCancellable *cancellable, GError **error)
+{
+  gboolean ret = FALSE;
+  GOptionContext *context;
+  glnx_unref_object OstreeSysroot *sysroot = NULL;
+  glnx_unref_object OstreeRepo *repo = NULL;
+  g_autoptr(GPtrArray) new_deployments = NULL;
+  glnx_unref_object OstreeDeployment *merge_deployment = NULL;
+  OstreeDeployment *booted_deployment = NULL;
+  OstreeDeployment *target_deployment = NULL;
+  g_autofree char *revision = NULL;
+  int sysroot_fd = -1;
+  g_autofree char *deployment_path = NULL;
+  glnx_fd_close int deployment_dfd = -1;
+  const char *deployment_mode = opt_hotfix ? "hotfix" : "development";
+
+  context = g_option_context_new ("Make the current deployment mutable (as a hotfix or development)");
+
+  if (!ostree_admin_option_context_parse (context, options, &argc, &argv,
+                                          OSTREE_ADMIN_BUILTIN_FLAG_SUPERUSER,
+                                          &sysroot, cancellable, error))
+    goto out;
+  
+  if (argc > 1)
+    {
+      ot_util_usage_error (context, "This command takes no extra arguments", error);
+      goto out;
+    }
+
+  if (!ostree_sysroot_load (sysroot, cancellable, error))
+    goto out;
+
+  sysroot_fd = ostree_sysroot_get_fd (sysroot);
+
+  /* Find the currently booted deployment, if any; we will ensure it
+   * is present in the new deployment list.
+   */
+  if (!ot_admin_require_booted_deployment_or_osname (sysroot, opt_osname,
+                                                     cancellable, error))
+    {
+      g_prefix_error (error, "Looking for booted deployment: ");
+      goto out;
+    }
+  
+  merge_deployment = ostree_sysroot_get_merge_deployment (sysroot, opt_osname);
+  if (!merge_deployment)
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED, "No previous deployment to duplicate");
+      goto out;
+    }
+  
+  booted_deployment = ostree_sysroot_get_booted_deployment (sysroot);
+  if (!booted_deployment)
+    {
+      if (!opt_fuse)
```

Comment:
I don't follow the logic here. We're only allowed to not be booted in the deployment if we're using fuse?